src/cudamatrix/cu-sparse-matrix.cc

```diff
@@ -70,7 +71,8 @@ template <typename Real>
 Real CuSparseMatrix<Real>::FrobeniusNorm() const {
 #if HAVE_CUDA == 1
   if (CuDevice::Instantiate().Enabled()) {
-    CuVector<Real> element_vec(*this);
+    CuVector<Real> element_vec(this->NumElements(), kUndefined);
+    this->CopyElementsToVec(&element_vec);
     return element_vec.Norm(2);
```

Comment:
We are computing sqrt(sum_{i=1}^N (x_i^2)) as the Frobenius norm right? Why
would the number of elements matter? Perhaps I misunderstood
CuVector.Norm()?

Guoguo

On Tue, Jan 5, 2016 at 3:31 PM, Daniel Povey <notifications@github.com>
wrote:

> In src/cudamatrix/cu-sparse-matrix.cc
> <https://github.com/kaldi-asr/kaldi/pull/434#discussion_r48911401>:
>
> > @@ -70,7 +71,8 @@ template <typename Real>
> >  Real CuSparseMatrix<Real>::FrobeniusNorm() const {
> >  #if HAVE_CUDA == 1
> >    if (CuDevice::Instantiate().Enabled()) {
> > -    CuVector<Real> element_vec(*this);
> > +    CuVector<Real> element_vec(this->NumElements(), kUndefined);
> > +    this->CopyElementsToVec(&element_vec);
> >      return element_vec.Norm(2);
>
> This code is not right, you need account for the fact that the matrix's
> size is different from the number of nonzero elements. Otherwise it's not
> really the frobenius norm.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/434/files#r48911401>.
>